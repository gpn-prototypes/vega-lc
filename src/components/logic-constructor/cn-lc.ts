import bemCn from 'bem-cn';

export const cnLogicConstructor = bemCn('LogicConstructorWidget');
