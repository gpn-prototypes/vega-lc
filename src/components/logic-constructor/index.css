.LogicConstructorWidget {
  position: relative;

  height: 100%;
}
