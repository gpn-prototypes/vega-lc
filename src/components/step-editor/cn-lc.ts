import bemCn from 'bem-cn';

export const cnStepEditor = bemCn('StepEditor');
