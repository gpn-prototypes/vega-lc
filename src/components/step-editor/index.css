:root {
  --step-editor-content-height: calc(100% - 50px);
  --step-editor-select-top-indent: 12px;
}

.StepEditor {
  position: fixed;
  top: 70px;
  right: 30px;

  width: 528px;
  height: calc(100vh - 100px);

  &__Tabs {
    position: relative;

    margin-bottom: 2px;
    padding-left: 12px;

    height: 35px;

    background-color: var(--color-bg-default);

    & > .Tabs {
      position: absolute;

      top: 2px;
    }
  }

  &__Content {
    padding: 24px var(--step-editor-select-top-indent);

    width: 100%;
    height: var(--step-editor-content-height);

    background-color: var(--color-bg-default);
  }

  &__ChanceSelect {
    width: 100%;
  }

  &__SelectLabel {
    margin: 0 0 8px;

    display: flex;
    align-items: center;

    font-size: 14px;
    line-height: 120%;

    color: var(--color-typo-secondary);
  }

  &__TimeFrames {
    margin-top: var(--step-editor-select-top-indent);

    width: 100%;
  }

  &__ExtremePoints {
    display: flex;

    width: 100%;

    & > div {
      width: 50%;
    }

    & > div:first-of-type {
      margin-right: 12px;
    }
  }

  &__DatePointSelect {
    width: 100%;
  }

  &__Continuous {
    margin-top: var(--step-editor-select-top-indent);
  }

  &__ContinuousSelect {
    width: calc(50% - 6px);
  }

  &__Cost {
    margin-top: var(--step-editor-select-top-indent);
  }

  &__CostSelect {
    width: calc(50% - 6px);
  }

  &__ParameterAppender {
    margin-top: calc(var(--step-editor-select-top-indent) * 2);
  }
}
