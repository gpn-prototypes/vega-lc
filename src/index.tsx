import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { ApolloProvider } from '@apollo/client';
import { Root } from '@gpn-prototypes/vega-root';

import getStore from './redux-store/index';
import { App } from './App';
import client from './client';

import './App.css';

ReactDOM.render(
  <Root initialPortals={[{ name: 'modalRoot' }]} defaultTheme="dark">
    <ApolloProvider client={client}>
      <Provider store={getStore()}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    </ApolloProvider>
  </Root>,
  document.getElementById('root'),
);
