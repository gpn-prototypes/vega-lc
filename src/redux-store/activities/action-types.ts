import { applyActionTypesNames } from '@/utils/apply-action-types-names';

export type ActivitiesActions =
  | 'SET_ACTIVITIES_LIST'
  | 'SET_SEARCH_STRING'
  | 'SET_IS_AUTO_FOCUS'
  | 'SET_ACTIVITIES_REF'
  | 'SET_DRAGGING_ELEMENTS'
  | 'SET_IS_DROPPING_ON_EXISTING_STEP';

const ActivitiesActionTypes: Record<ActivitiesActions, string> = {
  SET_ACTIVITIES_LIST: '',
  SET_SEARCH_STRING: '',
  SET_IS_AUTO_FOCUS: '',
  SET_ACTIVITIES_REF: '',
  SET_DRAGGING_ELEMENTS: '',
  SET_IS_DROPPING_ON_EXISTING_STEP: '',
};

applyActionTypesNames(ActivitiesActionTypes, 'Activities');

export { ActivitiesActionTypes };
