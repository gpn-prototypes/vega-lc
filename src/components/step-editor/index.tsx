import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { IconClose, Tabs } from '@gpn-prototypes/vega-ui';

import { cnStepEditor } from './cn-lc';
import CommonInfo from './CommonInfo';
import Conditions from './Conditions';
import Effects from './Effects';

import './index.css';

import { toggleStepEditor } from '@/redux-store/logic-constructor/actions';
import { getCanvasViewRef } from '@/redux-store/logic-constructor/selectors';

type Item = {
  name: string;
};

const StepEditor: React.FC = () => {
  const dispatch = useDispatch();
  const canvasViewRef = useSelector(getCanvasViewRef);

  const handleClose = (): void => {
    dispatch(toggleStepEditor(false));

    canvasViewRef?.current?.updateState({ selectedData: null });
  };

  const tabs = [
    {
      name: 'Общая информация',
    },
    {
      name: 'Условия начала и завершения',
    },
    {
      name: 'Эффекты',
    },
  ];

  const [val, setValue] = useState<Item | null>(tabs[0]);

  const renderContent = (): React.ReactElement => {
    switch (val?.name) {
      case tabs[1].name:
        return <Conditions />;
      case tabs[2].name:
        return <Effects />;
      default:
        return <CommonInfo />;
    }
  };

  return (
    <div className={cnStepEditor()}>
      <div className={cnStepEditor('Tabs')}>
        <Tabs
          size="s"
          view="clear"
          items={tabs}
          value={val}
          getLabel={(item): string => item.name}
          onChange={({ value }): void => setValue(value)}
        />
      </div>

      <div className={cnStepEditor('Content')}>
        {renderContent()}

        <IconClose size="s" className={cnStepEditor('CloseBtn').toString()} onClick={handleClose} />
      </div>
    </div>
  );
};

export default StepEditor;
