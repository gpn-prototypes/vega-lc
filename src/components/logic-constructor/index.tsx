import React from 'react';
import { useSelector } from 'react-redux';

import { ActivitiesWidget } from '../activities';
import { CanvasWidget } from '../canvas';
import StepEditor from '../step-editor';

import { cnLogicConstructor } from './cn-lc';

import './index.css';

import { InProgress } from '@/components/in-progress';

import { getIsStepEditorOpened } from '@/redux-store/logic-constructor/selectors';

export const LogicConstructorWidget: React.FC = () => {
  const isStepEditorOpened = useSelector(getIsStepEditorOpened);

  return (
    <div className={cnLogicConstructor()}>
      <ActivitiesWidget />
      <CanvasWidget />

      {isStepEditorOpened && <StepEditor />}
      <InProgress />
    </div>
  );
};
