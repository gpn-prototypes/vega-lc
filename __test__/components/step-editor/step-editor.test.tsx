import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { Store } from 'redux';
import configureMockStore from 'redux-mock-store';
import thunkMiddleware from 'redux-thunk';
import ResizeObserver from 'resize-observer-polyfill';

import '../../../src/types/global';

import StepEditor from '../../../src/components/step-editor';
import { cnStepEditor } from '../../../src/components/step-editor/cn-lc';
import { store } from '../../../src/redux-store';
import { LogicConstructorActionTypes } from '../../../src/redux-store/logic-constructor/action-types';
import initialState from '../../../src/redux-store/logic-constructor/initial-state';

jest.mock('../../../src/components/step-editor/CommonInfo', () => {
  return () => {
    return <div data-testid="common-info" />;
  };
});

jest.mock('../../../src/components/step-editor/Conditions', () => {
  return () => {
    return <div data-testid="conditions" />;
  };
});

jest.mock('../../../src/components/step-editor/Effects', () => {
  return () => {
    return <div data-testid="effects" />;
  };
});

const middlewares = [thunkMiddleware];
const mockStore = configureMockStore(middlewares);

const renderComponent = (customStore: Store) => {
  return render(
    <Provider store={customStore}>
      <StepEditor />
    </Provider>,
  );
};

describe('Step Editor', () => {
  test('рендерится без ошибок', () => {
    window.ResizeObserver = ResizeObserver;

    const dom = renderer
      .create(
        <Provider store={store}>
          <StepEditor />
        </Provider>,
      )
      .toJSON();
    expect(dom).toMatchSnapshot();
  });

  test('при переключении вкладки отображается нужный компонент', async () => {
    const mockedStore = mockStore({ logicConstructor: initialState });

    const component = await renderComponent(mockedStore);

    const tabs = component.container.querySelectorAll('.TabsTab');

    if (tabs[1]) {
      fireEvent.click(tabs[1]);
    }

    const conditions = component.queryByTestId('conditions');
    await waitFor(() => expect(conditions).toBeInTheDocument());

    if (tabs[2]) {
      fireEvent.click(tabs[2]);
    }

    const effects = component.queryByTestId('effects');
    await waitFor(() => expect(effects).toBeInTheDocument());

    if (tabs[0]) {
      fireEvent.click(tabs[0]);
    }

    const commonInfo = component.queryByTestId('common-info');
    await waitFor(() => expect(commonInfo).toBeInTheDocument());
  });

  test('редактор шага закрывается по нажатию на кнопку', async () => {
    const updateState = jest.fn();
    const mockCanvasViewRef = { current: { updateState } };
    const mockedStore = mockStore({
      logicConstructor: {
        ...initialState,
        isStepEditorOpened: true,
        canvasViewRef: mockCanvasViewRef,
      },
    });

    const component = await renderComponent(mockedStore);

    const closeButton = component.container.querySelector(`.${cnStepEditor('CloseBtn')}`);

    if (closeButton) {
      fireEvent.click(closeButton);
    }

    const expectedAction = {
      type: LogicConstructorActionTypes.TOGGLE_STEP_EDITOR,
      isStepEditorOpened: false,
    };

    await waitFor(() => expect(mockedStore.getActions()).toContainEqual(expectedAction));
    expect(updateState).toHaveBeenCalledTimes(1);
  });
});
