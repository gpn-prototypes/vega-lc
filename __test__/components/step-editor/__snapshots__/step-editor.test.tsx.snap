// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step Editor рендерится без ошибок 1`] = `
<div
  className="StepEditor"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        style={
          Object {
            "overflowX": "hidden",
          }
        }
      >
        <div
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="Tabs Tabs_size_s Tabs_view_clear VegaTabs__Native"
            style={
              Object {
                "width": 0,
              }
            }
          >
            <div
              className="Tabs-List"
            >
              <button
                className="TabsTab TabsTab_checked MixFocus MixFocus_before Tabs-Tab"
                role="tab"
                type="button"
              >
                Общая информация
              </button>
              <button
                className="TabsTab MixFocus MixFocus_before Tabs-Tab"
                onClick={[Function]}
                role="tab"
                type="button"
              >
                Условия начала и завершения
              </button>
              <button
                className="TabsTab MixFocus MixFocus_before Tabs-Tab"
                onClick={[Function]}
                role="tab"
                type="button"
              >
                Эффекты
              </button>
            </div>
            <div
              className="Tabs-WrapperRunningLine"
            >
              <div
                className="Tabs-RunningLine Tabs-RunningLine_mounted"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        <button
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
          onClick={[Function]}
          type="button"
        >
          <span
            className="Icon Icon_size_m IconArrowRight"
          >
            <svg
              className="Icon-Svg"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <div
      data-testid="common-info"
    />
    <span
      className="Icon Icon_size_s IconClose StepEditor__CloseBtn"
      onClick={[Function]}
    >
      <svg
        className="Icon-Svg"
        viewBox="0 0 16 16"
      >
        <path
          clipRule="evenodd"
          d="M3.757 2.343L2.343 3.757 6.586 8l-4.243 4.243 1.414 1.414L8 9.414l4.243 4.243 1.414-1.414L9.414 8l4.243-4.243-1.414-1.414L8 6.586 3.757 2.343z"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </div>
</div>
`;
